test/unittest/unittest_nntrainer_tensor_neon_fp16.cpp

```diff
@@ -150,11 +150,54 @@ TEST(nntrainer_Tensor, l2norm) {
   result_fp32 = input_fp32.l2norm();
 
   // absolute error
-  const float epsilon = 1e-2;
+  const float epsilon = 1e-3 * width;
 
   EXPECT_NEAR(result_neon, result_fp32, epsilon);
 }
 
+TEST(nntrainer_Tensor, l2norm_big768) {
+
+  nntrainer::TensorDim::TensorType t_type_nchw_fp16 = {
+    nntrainer::Tformat::NCHW, nntrainer::Tdatatype::FP16};
+
+  nntrainer::TensorDim::TensorType t_type_nchw_fp32 = {
+    nntrainer::Tformat::NCHW, nntrainer::Tdatatype::FP32};
+
+  size_t batch = 1;
+  size_t channel = 1;
+  size_t height = 768;
+  size_t width = 768;
+
+  nntrainer::Tensor input(
+    nntrainer::TensorDim(1, 1, height, width, t_type_nchw_fp16));
+
+  nntrainer::Tensor input_fp32(
+    nntrainer::TensorDim(1, 1, height, width, t_type_nchw_fp32));
+
+  const float alpha = 1e-1;
+  const int MOD = 10;
+
+  GEN_TEST_INPUT(input, ((i * j * (batch * height * channel) +
+                          j * (batch * height) + k * (width) + l + 1) %
+                         MOD) *
+                          alpha);
+  GEN_TEST_INPUT(input_fp32, ((i * j * (batch * height * channel) +
+                               j * (batch * height) + k * (width) + l + 1) %
+                              MOD) *
+                               alpha);
+
+  __fp16 result_neon;
+  float result_fp32;
+
+  result_neon = input.l2norm();
+  result_fp32 = input_fp32.l2norm();
+
+  float ErrorNeon = abs(result_neon - result_fp32);
+
+  const float epsilon = 1e-3 * width;
```

Comment:
what's the reasoning for multiplying width? is it to sum errors of num of width?

Reply:
There is unavoidable round-off error that linearly increase w.r.t. tensor size in half-precision computation. Meanwhile, I am comparing half-precision result with fp32 result. If they were using the same precision, I could use mathematically-defined criteria, but it is not. So I have been using this kind of epsilon to compensate such error.
Strictly speaking, this function makes (M x N) Matrix to one single value, so something like:
```C++
  const float epsilon = 1e-3 * height * width;
  // or const float epsilon = sqrt(1e-3 * height * width); ???
``` 
would be better in terms of consistency, but I just left it this way this way for stricter condition.

Reply:
makes sense. if so, I think comparing fp16 with fp32 could be pointless since it would never be close (especially when the dimension increases). how about we compare the fp16 tensor l2 norm value with other frameworks such as PyTorch or Tensorflow?

Reply:
Nice approach...
Then, is there a efficient way to compare Pytorch output and ourcase in unittest codes?
I don't really prefer hard-coded unittest cases like we do now...

Reply:
I tested the added test case, `l2norm_big768`, with PyTorch and Tensorflow. The L2 norm value for the float32 tensor matches with both frameworks. However, there's a mismatch for float16.

#### NNTrainer
Before: 45.25
After: 410.25

#### PyTorch
156.7500

#### Tensorflow
inf

This fix gets the closest value to the float32 tensor, which is 409.999, so there's no need to compare with other frameworks. I think it is good to go :)

Reply:
Based on the output from PyTorch / Tensorflow, I think they are just computing the value with full-fp16 precision.
Mathematically it is not really a wrong value, but it is not actually desired way in our framework I guess.
Anyway, thanks for pointing this issue out! 👍